qiskit/dagcircuit/dagdependency_v2.py

```diff
@@ -0,0 +1,740 @@
+                self.cindices_map[new_node] = [self.find_bit(c).index for c in cond_bits]
+            else:
+                self.cindices_map[new_node] = []
+        else:
+            self.qindices_map[new_node] = []
+            self.cindices_map[new_node] = []
+
+    def _update_edges(self):
+        """
+        Updates DagDependencyV2 by adding edges to the newly added node (max_node)
+        from the previously added nodes.
+        For each previously added node (prev_node), an edge from prev_node to max_node
+        is added if max_node is "reachable" from prev_node (this means that the two
+        nodes can be made adjacent by commuting them with other nodes), but the two nodes
+        themselves do not commute.
+
+        Currently. this function is only used when creating a new DAGDependencyV2 from another
+        representation of a circuit, and hence there are no removed nodes (this is why
+        iterating over all nodes is fine).
+        """
+        max_node_id = len(self._multi_graph) - 1
+        max_node = self.get_node(max_node_id)
+
+        reachable = [True] * max_node_id
+
+        # Analyze nodes in the reverse topological order.
+        # An improvement to the original algorithm is to consider only direct predecessors
+        # and to avoid constructing the lists of forward and backward reachable predecessors
+        # for every node when not required.
+        for prev_node_id in range(max_node_id - 1, -1, -1):
+            if reachable[prev_node_id]:
+                prev_node = self.get_node(prev_node_id)
+
+                if not self.comm_checker.commute(
+                    prev_node.op,
+                    prev_node.qargs,
+                    prev_node.cargs,
+                    max_node.op,
+                    max_node.qargs,
+                    max_node.cargs,
+                ):
+                    # If prev_node and max_node do not commute, then we add an edge
+                    # between the two, and mark all direct predecessors of prev_node
+                    # as not reaching max_node.
+                    self._multi_graph.add_edge(prev_node_id, max_node_id, {"commute": False})
+
+                    predecessor_ids = self.predecessor_indices(prev_node_id)
+                    for predecessor_id in predecessor_ids:
+                        reachable[predecessor_id] = False
+            else:
+                # If prev_node cannot reach max_node, then none of its predecessors can
+                # reach max_node either.
+                predecessor_ids = self.predecessor_indices(prev_node_id)
+                for predecessor_id in predecessor_ids:
+                    reachable[predecessor_id] = False
+
+    def get_node(self, node_id):
+        """
+        Args:
+            node_id (int): label of considered node.
+
+        Returns:
+            node: corresponding to the label.
+        """
+        return self._multi_graph.get_node_data(node_id)
```

Comment:
This is needed in template matching so not important there, but also in `_update_edges` since that method is written using node indices instead of node object ids. I did a first try to convert this method to using object ids, but could not come up with an efficient way to do it. So for now I would leave this and address it later.

The use of node indices could also be a problem if nodes are inserted or removed and then iterated over.

Reply:
Leaving this in until I rewrite `_update_edges` to not use indices. Also used in some of the tests, which will need to be redone later.

Reply:
Agreed. What about making `get_node` into an internal function (i.e. just renaming it to `_get_node`), so that we will able to freely change/remove it later?

Reply:
BTW, let's rewrite `_update_edges` in a follow-up PR, not now.

Reply:
Agreed on `_get_node` and redo `_update_edges` later.